app/lib/actions/katello/repository/destroy.rb

```diff
@@ -70,11 +70,7 @@ def handle_acs_product_removal(repository)
           # Remove products from ACS's that contain no repositories which both
           # match the ACS content type and have a non-nil URL
           product = repository.product
-          repo_content_types = Set.new
-          product.repositories.each do |test_repo|
-            # we need to check id because test_repo will still contain the old, non-nil url
-            repo_content_types.add(test_repo.content_type) if (repository.id != test_repo.id) && test_repo.url.present?
-          end
+          repo_content_types = ::Katello::RootRepository.where(:id => product.repositories.where.not(:id => repository.id).select(:root_id)).where.not(:url => [nil, '']).distinct.pluck(:content_type)
```

Comment:
@qcjames53 might have some more context, it was a long time ago that this code was written though :)

Reply:
@ianballou Thanks for bringing me in on this. Good catch with the possibility of two repositories sharing the same root in the test case. It looks like `url_sum` always comes out to 1 with the test's data due to some wacky ruby shenanigans with referencing.

Another issue I found with some byebugging is that the root urls are never actually updated if you're coming at the test from the `::Katello::RootRepository` side of things, so the urls the destroy function tests against are still the old values. This isn't an issue in normal use, just a side effect of how I set up the test.

Here's a corrected version of the test. I luckily was able to repurpose the rhel_7 repo since it is the only repo pointing to the rhel_7 root repository:
```
  class DestroyTest < TestBase
...
    let(:published_rhel7_repository) { katello_repositories(:rhel_7_no_arch) }
...
    it 'plans ACS product removal when deleting the last repo with URL' do
      ::Katello::SmartProxyAlternateContentSource.create!(alternate_content_source_id: simplified_acs.id, smart_proxy_id: proxy.id, repository_id: published_rhel7_repository.id)
      action = create_action action_class
      action.stubs(:action_subject).with(published_rhel7_repository)

      # manually remove the URLs from all repos in product except repository
      testing_repo = repository.product.repositories[4]
      testing_repo.product.repositories.each do |repo|
        repo.root.url = nil unless repo.id == testing_repo.id || repo.root.id == testing_repo.root.id
        repo.root.save!(validate: false)
      end
      url_sum = testing_repo.product.repositories.count do |repo|
        repo.root.url.present?
      end
      assert_equal(1, url_sum) # double check there's only one URL left

      # Since there is only one URL remaining, the product should be removed
      plan_action action, published_rhel7_repository, remove_from_content_view_versions: true
      simplified_acs.reload
      assert_equal(0, simplified_acs.products.length)
    end
  end
```